sdk/internal/perf/perf.go

```diff
@@ -0,0 +1,329 @@
+		}
+	}
+
+	if WarmUp > 0 {
+		warmUpStart := time.Now()
+		for time.Since(warmUpStart).Seconds() < float64(WarmUp) {
+			err := p.Run(context.Background())
+			if err != nil {
+				c <- runResult{err: err}
+			}
+		}
+	}
+
+	timeStart := time.Now()
+	totalCount := 0
+	lastPrint := 1.0
+	perSecondCount := make([]int, 0)
+	w := tabwriter.NewWriter(os.Stdout, 16, 8, 1, ' ', tabwriter.AlignRight|tabwriter.Debug)
+	for time.Since(timeStart).Seconds() < float64(Duration) {
+		err := p.Run(context.Background())
+		if err != nil {
+			c <- runResult{err: err}
+		}
+		totalCount += 1
+
+		// Every second (roughly) we print out an update
+		if time.Since(timeStart).Seconds() > float64(lastPrint) {
+			thisCount := totalCount - sumInts(perSecondCount)
+			perSecondCount = append(perSecondCount, thisCount)
+			_, err = fmt.Fprintf(
+				w,
+				"%s\t%s\t%.2f\t\n",
+				commaIze(thisCount),
+				commaIze(totalCount),
+				float64(sumInts(perSecondCount))/time.Since(timeStart).Seconds(),
+			)
+			if err != nil {
+				c <- runResult{err: err}
+			}
+			lastPrint = time.Since(timeStart).Seconds() + 1.0
+			w.Flush()
+		}
+	}
+
+	elapsed := time.Since(timeStart).Seconds()
+
+	// Stop the proxy now
+	err := stop(p.GetMetadata().Name, nil)
+	if err != nil {
+		c <- runResult{err: err}
+	}
+	setRecordingMode("live")
+	c <- runResult{count: totalCount, timeInSeconds: elapsed, err: nil}
+}
+
+// runCleanup takes care of the semantics for tearing down a single iteration of a performance test.
+func runCleanup(p PerfTest) error {
+	err := p.Cleanup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func runGlobalCleanup(p PerfTest) error {
+	err := p.GlobalCleanup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type runResult struct {
+	count         int
+	timeInSeconds float64
+	err           error
+}
+
+func runPerfTest(p NewPerfTest) error {
+	err := runGlobalSetup(p(nil))
+	if err != nil {
+		panic(err)
+	}
+
+	var channels []chan runResult
+	var perfTests []PerfTest
+
+	fmt.Println("=== Test ===")
+
+	w := tabwriter.NewWriter(os.Stdout, 16, 8, 1, ' ', tabwriter.AlignRight)
+	fmt.Fprintln(w, "Current\tTotal\tAverage\t")
+	err = w.Flush()
+	if err != nil {
+		panic(err)
+	}
+
+	for idx := 0; idx < Parallel; idx++ {
+		options := &PerfTestOptions{}
+		if TestProxy != "" {
+			transporter, err := NewProxyTransport(&TransportOptions{TestName: p(nil).GetMetadata().Name})
+			if err != nil {
+				panic(err)
+			}
+			options.ProxyInstance = transporter
+		} else {
+			options.ProxyInstance = defaultHTTPClient
+		}
+		options.ParallelIndex = idx
+
+		perfTest := p(options)
+		perfTests = append(perfTests, perfTest)
+
+		// Run the setup for a single instance
+		err := runSetup(perfTest)
+		if err != nil {
+			return err
+		}
+
+		// Create a thread for running a single test
+		wg.Add(1)
+		c := make(chan runResult, 1) // Create a buffered channel
+		channels = append(channels, c)
+		go runTest(perfTest, c)
+	}
+
```

Comment:
I'm seeing an issue at the end of the duration, where Linux hangs and Windows throws with this error:

```
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0x0)
        C:/Program Files/Go/src/runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0x6a3a00)
        C:/Program Files/Go/src/sync/waitgroup.go:130 +0x71
github.com/Azure/azure-sdk-for-go/sdk/internal/perf.runPerfTest(0x706908)
        D:/Git/go/sdk/internal/perf/perf.go:230 +0x557
github.com/Azure/azure-sdk-for-go/sdk/internal/perf.Run({0xc000113f60, 0x0, 0x2})
        D:/Git/go/sdk/internal/perf/perf.go:325 +0x358
main.main()
        D:/Git/go/sdk/internal/perf/testdata/perf/main.go:6 +0x45

goroutine 8 [chan send]:
github.com/Azure/azure-sdk-for-go/sdk/internal/perf.runTest({0x74ffd8, 0xc00015a6c0}, 0x0)
        D:/Git/go/sdk/internal/perf/perf.go:157 +0xbc5
created by github.com/Azure/azure-sdk-for-go/sdk/internal/perf.runPerfTest
        D:/Git/go/sdk/internal/perf/perf.go:227 +0x16c

goroutine 9 [chan send]:
github.com/Azure/azure-sdk-for-go/sdk/internal/perf.runTest({0x74ffd8, 0xc00015a720}, 0x0)
        D:/Git/go/sdk/internal/perf/perf.go:157 +0xbc5
created by github.com/Azure/azure-sdk-for-go/sdk/internal/perf.runPerfTest
        D:/Git/go/sdk/internal/perf/perf.go:227 +0x16c
exit status 2
```